__init__.py

```diff
@@ -0,0 +1,333 @@
+"""__init__.py.
+
+Last Update: May 29 2024
+"""
+
+import re
+from collections.abc import Iterator
+from dataclasses import dataclass
+from typing import Callable, Iterable, List, Union
+
+import spacy
+from spacy.tokens.doc import Doc
+from timer import timer
+
+from rollingwindows import helpers
+from rollingwindows.registry import rw_components
+
+
+def get_rw_component(id: str):
+    """Get a component from the registry by id.
+
+    Args:
+        id (str): The registry id of the component
+
+    Returns:
+        The component class.
+    """
+    return rw_components.get(id)
+
+
+def sliding_str_windows(
+    input: Union[List[spacy.tokens.span.Span], spacy.tokens.doc.Doc, str],
+    n: int = 1000,
+    alignment_mode: str = "contract",
+) -> Iterator:
+    """Return a generator of string windows.
+
+    Args:
+        input (Union[List[spacy.tokens.span.Span], spacy.tokens.doc.Doc, str]): A spaCy doc or a list of spaCy spans.
+        n (int): The size of the window.
+        window_units (str): The type of units to use ("characters", "tokens", "lines", "sentences", "spans").
+        alignment_mode (str): How character indices snap to token boundaries.
+        - "strict" (no snapping)
+        - "contract" (span of all tokens completely within the character span)
+        - "expand" (span of all tokens at least partially covered by the character span)
+
+    Returns:
+        A generator of window strings.
+
+    Note:
+        Window boundaries are snapped to token boundaries in the original doc.
+        "Contract" means that the window will contain all tokens completely
+        within the boundaries of `i:i + n`. "Expand" means that window will
+        contain all tokens partially withn those boundaries. Setting
+        `alignment_mode="strict"` in `doc.char_span()` is not advised
+        because it returns `None` for any string that cannot be aligned to
+        token boundaries. As a result, a slice method is used if you want
+        to simply cut the text strictly on `n` characters.
+    """
+    # TODO: We have to iterate through the input twice to get the boundaries.
+    if isinstance(input, list):
+        input_spans = [span.as_doc() for span in input]
+        boundaries = [(i, i + n) for i in range(len(input_spans))]
+        for start, end in boundaries:
+            yield Doc.from_docs(input_spans[start:end]).text.strip()
+    else:
```

Comment:
Could you change to using a batched generator to make this more memory efficient for really long texts? You could then use parallel processing to create the windows?

Reply:
Is there a place where you could cache the window calculations?